SE-2672 Added Tox Tests and Updated to Version 0.2

nizarmah opened the pull request:
Adds tox tests for the following _envlist_:
```
py{35}-django{22}, py{38}-django{22,30}, py38-quality
```

**JIRA tickets**: SE-2672

**Merge deadline**: None

**Testing instructions**:

1. Run _tox_ tests using `tox`.

**Reviewers**
- [ ] @pomegranited

nizarmah committed:
Added Tox Unit Tests to Check For Content Type Key in Exclude Dictionary

Added LmsSearchFilterGenerator mock class to avoid creating a django app. Creating a django app was resulting in many dependencies / django apps. In addition, it was resulting in an error. It was more convenient to unit test the ExcludeCapaSearchFilterGenerator with a mock LmsSearchFilterGenerator mock class.

The LmsSearchFilterGenerator and the SearchFilterGenerator already have unit tests, thus, the unit test only for ExcludeCapaSearchFilterGenerator.

nizarmah committed:
Added PylintRc from EdxLint and Updated Files to Fix Pylint Warnings

nizarmah committed:
Updated Setup to Version 0.2 and Removed Python 2.7 Support

nizarmah committed:
Updated README to Add Testing Instruction

pomegranited commented on opencraft_courseware_search_filters/exclude_capa.py:
```diff
@@ -16,7 +19,7 @@ def exclude_dictionary(self, **kwargs):
         Exclude CAPA results
         """
-        exclude_dictionary = super(LmsSearchFilterGenerator, self).exclude_dictionary(
+        exclude_dictionary = super(LmsSearchFilterGenerator, self).exclude_dictionary(  # pylint: disable=bad-super-call
```
Having to disable [bad-super-call](https://docs.pylint.org/en/1.6.0/features.html#id33) looks like a bug in the previous code?

```suggestion
        exclude_dictionary = super().exclude_dictionary(
```

nizarmah replied:
It might have been a bug, but the thing is that the `exclude_dictionary` in the `LmsSearchFilterGenerator` and in the parent (`SearchFilterGenerator`) are really different. That's why I disabled the `bad-super-call`.

### SearchFilterGenerator
```
def exclude_dictionary(self, **kwargs):
    """ base implementation which excludes nothing """
    return {}
```

Meanwhile, 

### LmsSearchFilterGenerator
```
def exclude_dictionary(self, **kwargs):
    """
        Exclude any courses defined outside the current org.
    """
    exclude_dictionary = super(LmsSearchFilterGenerator, self).exclude_dictionary(**kwargs)
    course_org_filter = configuration_helpers.get_current_site_orgs()
    # If we have a course filter we are ensuring that we only get those courses above
    if not course_org_filter:
        org_filter_out_set = configuration_helpers.get_all_orgs()
        if org_filter_out_set:
            exclude_dictionary['org'] = list(org_filter_out_set)

    return exclude_dictionary
```

Do you think Samuel's input on this could help?

pomegranited replied:
Sure.. @swalladge  Do you remember why you did the inheritance this way?

samuelallan72 replied:
ummm.. no idea. Did the way I call super cause it to inherit from SearchFilterGenerater.exclude_dictionary instead of LmsSearchFilterGenerator.exclude_dictionary? If so, then yep it's a bug.

Since we're definitely on python3 now, probably best to avoid the gotcha and just call `super().exclude_dictionary...`?

samuelallan72 replied:
I probably copy-pasted and forgot to update the class in the super() call... :facepalm:

nizarmah replied:
Thanks for the quick reply @swalladge :smile:

nizarmah replied:
@pomegranited I'll go for the `super()` as @swalladge suggested. I updated your suggestion, however, I realized that I also need to update the Test. So I'll push those as a single commit.

pomegranited commented on tests/unit/test_exclude_capa.py:
```diff
@@ -0,0 +1,17 @@
+class TestExcludeCapaSearchFilterGenerator(unittest.TestCase):
+    def test_exclude_dictionary_for_content_type_key(self):
+        search_filter = ExcludeCapaSearchFilterGenerator()
+        self.assertIn("content_type", search_filter.exclude_dictionary())
```
Since `ExcludeCapaSearchFilterGenerator`'s job is to exclude CAPA content types, you can even go this far:

```suggestion
        self.assertEqual("CAPA", search_filter.exclude_dictionary()["content_type"])
```

nizarmah replied:
Oh! That's actually a much better assertion, I should've paid attention to that. Thanks :smile: :+1:

pomegranited commented on tests/unit/test_exclude_capa.py:
```diff
@@ -0,0 +1,17 @@
+"""
+import unittest
+
+from opencraft_courseware_search_filters import ExcludeCapaSearchFilterGenerator
```
@nizarmah  You mentioned in the ticket that you mocked `lms.lib.courseware_search.lms_filter_generator.LmsSearchFilterGenerator` to get these tests to run outside the edxapp virtualenv, but I don't see that?

Did you miss a push here?

nizarmah replied:
No, not at all. So basically, the `ExcludeCapaSearchFilterGenerator` is importing the `lms.lib.courseware_search.lms_filter_generator`. Therefore, I have the `lms` folder, with the necessary path added to the `PYTHONPATH` with `tox`.

Therefore, whenever the `ExcludeCapaSearchFilterGenerator` imports the `LmsSearchFilterGenerator`, it is importing the "fake" / mock class that I created. The mock class is an actual class created instead of mocking imports with `unittest.mock` because I was having to deal with import module errors and I wasn't able to have the flexibility I wanted from the `unittest.mock`. Might have been my inexperience with it, but I was planning to partially mock the parent class without affecting the one inheriting it (the part about "not affecting the one inheriting it" is what I don't think is possible with `unittest.mock`)

nizarmah replied:
Should I mark this as resovled @pomegranited ?

pomegranited replied:
@nizarmah  Ah that's it!  There's no `tests/utils/lms/lib` dir added to this PR yet :)

nizarmah replied:
@pomegranited it says they are gitignored !? oh my, I didn't pay attention to that. I'll fix it and push. Sorry about that.

nizarmah replied:
Okay, fixed it. Phew. Sorry about that.

pomegranited commented on pylintrc:
```diff
@@ -0,0 +1,381 @@
+# ***************************
```
I don't think you'll need to add this whole file..  Would recommend trying pylint without adding a `pylintrc`;  Their [default settings](https://github.com/PyCQA/pylint/blob/master/pylintrc) are good.  You can just fix whatever pylint complains about, if anything. It's a tiny repo, so we're not too concerned about keeping letter-perfect with [edx-lint](http://github.com/edx/edx-lint).

pomegranited commented on tox.ini:
```diff
@@ -0,0 +1,28 @@
+    PYTHONPATH = {toxinidir}/tests/utils
+
+deps =
+    django22: Django==2.2.12
```
nit: [use juniper's django version](https://github.com/edx/edx-platform/blob/open-release/juniper.1/requirements/edx/django.txt):
```suggestion
    django22: Django==2.2.13
```

nizarmah replied:
Yeah, sorry about that. I should've paid attention to that.

pomegranited reviewed (changes requested):
Thanks for this @nizarmah !  Changes requested below.

nizarmah committed:
Updated Tox File to Use Juniper's Django Version

Co-authored-by: Jillian Vogel <jill@opencraft.com>

nizarmah committed:
Changed Content Type Assertion to Check For CAPA Content Type

Co-authored-by: Jillian Vogel <jill@opencraft.com>

nizarmah committed:
Used Default Settings for Pylint with Py38-Quality Test

Had to disable 'too-few-public-methods- warning for ExcludeCapaSearchFilterGenerator

nizarmah committed:
Added Mock LmsSearchFilterGenerator After Updating GitIgnore

GitIgnore was ignoring all lib/ paths; therefore, added an exception for tests/utils/lms/lib

nizarmah committed:
Updated Super Call for Exclude Dictionary

Ended up removing the SearchFilterGenerator as it was no longer needed.

pomegranited reviewed (approved):
:+1: Works great, thank you @nizarmah !  Ready to squash, merge, and tag.

- [x] I tested this:
   * ran `tox` as instructed and verified tests pass and coverage 
   * Installed this update in my master devstack and [enabled courseware search indexing](https://edx.readthedocs.io/projects/edx-installing-configuring-and-running/en/latest/configuration/edx_search.html).
      Verified that with `SEARCH_FILTER_GENERATOR: 'opencraft_courseware_search_filters.ExcludeCapaSearchFilterGenerator'` set, searching for `chair` produces no search results in the demo course (this word is only in a CAPA problem).
      Verified that with the default `SEARCH_FILTER_GENERATOR`, searching for `chair` does return the CAPA problem.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] I made sure any change in configuration variables is reflected in the corresponding
    client's `configuration-secure` repository.
